drivers/video/hm01b0.c

```diff
@@ -261,10 +274,84 @@ static bool hm01b0_check_connection(const struct device *dev)
 	return (model_id == HM01B0_ID);
 }
 
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(pwdn_gpios) || DT_ANY_INST_HAS_PROP_STATUS_OKAY(reset_gpios)
+static bool hm01b0_try_reset_pwdn_pins(const struct device *dev, uint8_t iter)
+{
+	const struct hm01b0_config *config = dev->config;
+
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(pwdn_gpios)
+	gpio_pin_set_dt(&config->pwdn, iter >> 1);
+#else
+	if (iter >> 1) {
+		return false; /* cut iterations in half if not defined */
+	}
+#endif
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(reset_gpios)
+	gpio_pin_set_dt(&config->reset, iter & 1);
+#else
+	if (iter & 1) {
+		return false; /* cut iterations in half if not defined */
+	}
+#endif
+	/* lets try a couple of iterations before we punt */
+	uint8_t retry_count = 3;
+
+	while (retry_count) {
+		k_sleep(K_MSEC(10));
+		if (hm01b0_check_connection(dev)) {
+			LOG_INF("Reset/pwdn pins valid:%u", retry_count);
+			return true;
+		}
+		retry_count--;
+	}
+	return false;
+}
+#endif
+
 static int hm01b0_init(const struct device *dev)
 {
+	const struct hm01b0_config *config = dev->config;
 	int ret;
 
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(pwdn_gpios) || DT_ANY_INST_HAS_PROP_STATUS_OKAY(reset_gpios)
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(pwdn_gpios)
+	/* Power up camera module */
+	if (config->pwdn.port != NULL) {
+		if (!gpio_is_ready_dt(&config->pwdn)) {
+			return -ENODEV;
+		}
+		ret = gpio_pin_configure_dt(&config->pwdn, GPIO_OUTPUT_ACTIVE);
+		if (ret < 0) {
+			LOG_ERR("Could not clear power down pin: %d", ret);
+			return ret;
+		}
+	}
+#endif
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(reset_gpios)
+	/* Reset camera module */
+	if (config->reset.port != NULL) {
+		if (!gpio_is_ready_dt(&config->reset)) {
+			return -ENODEV;
+		}
+		ret = gpio_pin_configure_dt(&config->reset, GPIO_OUTPUT_ACTIVE);
+		if (ret < 0) {
+			LOG_ERR("Could not set reset pin: %d", ret);
+			return ret;
+		}
+	}
+#endif
+	bool found_id = false;
+
+	for (uint8_t iter = 0; iter < 4; iter++) {
+		found_id = hm01b0_try_reset_pwdn_pins(dev, iter);
+		if (found_id) {
+			break;
+		}
+	}
+#endif
+
+
+	LOG_INF("hm01b0_init check connection");
 	if (!hm01b0_check_connection(dev)) {
 		LOG_ERR("%s is not ready", dev->name);
 		return -ENODEV;
```

Comment:
If I understand correctly, here we have first try to configure reset/pwdn pin as needed, and then try to toggle them repeatedly a configurable number of times?

It seems like the latter is all what need to be supported as it would act as both?

```suggestion
	for (int i = 0; i < HM01B0_INIT_ATTEMPTS; i++) {
#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(reset_gpios)
		if (config->reset.port != NULL) {
			if (!gpio_is_ready_dt(&config->reset)) {
				return -ENODEV;
			}
			ret = gpio_pin_configure_dt(&config->reset, GPIO_OUTPUT_ACTIVE);
			if (ret < 0) {
				LOG_ERR("Could not set reset pin: %d", ret);
				return ret;
			}
		}
#endif

#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(reset_gpios)
		/* Reset camera module */
		if (config->reset.port != NULL) {
			if (!gpio_is_ready_dt(&config->reset)) {
				return -ENODEV;
			}
			ret = gpio_pin_configure_dt(&config->reset, GPIO_OUTPUT_ACTIVE);
			if (ret < 0) {
				LOG_ERR("Could not set reset pin: %d", ret);
				return ret;
			}
		}
#endif
	}

	if (!hm01b0_check_connection(dev)) {
		LOG_ERR("%s is not ready", dev->name);
		return -ENODEV;
	}
```

Did I understand the goal corectly?

Reply:
The problem with your possible replacement code, is that it  potentially can try all 4 combination of
pin setting as I mentioned, but if the first one does not succeed it will return (-ENODEV).

The current code Will fail in the call hm01b0_try_reset_pwdn_pinsm which does like hm01b0_check_connection,
except it will not bail at this point but continue to the next combination and try again.